### PR TITLE
Add drmSetMaster and drmDropMaster calls to FreeBSD direct session

### PR DESCRIPTION
**Pull request by sghctoma**
Hi all,

I have been using Sway on my HardenedBSD 12-CURRENT machine for a few months now, and I think maybe it's time to PR my patches. This one aims to fix a problem that is caused by FreeBSD not using major device numbers to indicate device type.
Because of this, there were no drmSetMaster and drmDropMaster calls, which caused some problems when switching TTYs, exiting the compositor, or running multiple compositors (or X server) at the same time. Using string comparison (device paths) to achieve this feels hacky, but I don't think there is a better solution.

### CODE REVIEW

**Comment by emersion** [comment]
Not a fan of this. Any way to patch FreeBSD to correctly use major device numbers?

File: backend/session/direct-freebsd.c [hunk 1]
```diff
@@ -79,6 +87,8 @@ static void direct_session_destroy(struct wlr_session *base) {
 	ioctl(session->tty_fd, KDSETMODE, KD_TEXT);
 	ioctl(session->tty_fd, VT_SETMODE, &mode);
 
+	ioctl(session->tty_fd, VT_ACTIVATE, 1);
```
[the remark this answers is not in the document]
Comment (sghctoma) [reply backend/session/direct-freebsd.c]: Yep, sorry for that, it was a temporary change I forgot to remove before committing. I've pushed a new commit that activates the VT the compositor was started from.

**Comment by sghctoma** [comment]
It's not that FreeBSD uses device numbers incorrectly, it just uses them differently. It used to use them the way Linux does, but that behaviour was dropped in FreeBSD 5, and now they are just internal identifiers. I think it was done to simplify defvs. All in all, this won't change.

I haven't thought of it until today, but maybe I could use some ioctls to determine device type. Like I could call DRM_IOCTL_VERSION and EVIOCGVERSION, and check for errors to determine if I'm speaking to a drm or input device. What do you think of this approach?

**Comment by SirCmpwn** [comment]
>I haven't thought of it until today, but maybe I could use some ioctls to determine device type. Like I could call DRM_IOCTL_VERSION and EVIOCGVERSION, and check for errors to determine if I'm speaking to a drm or input device. What do you think of this approach?

+1

**Comment by sghctoma** [comment]
This took a bit longer than expected (had to chase down a bug in the drm driver) but I have finally committed the ioctl-based device type discovery.

File: backend/session/direct-ipc.c [hunk 2]
```diff
@@ -193,8 +208,11 @@ int direct_ipc_open(int sock, const char *path) {
 
 	send_msg(sock, -1, &msg, sizeof(msg));
 
-	int fd, err;
-	recv_msg(sock, &fd, &err, sizeof(err));
+	int fd, err, ret;
+	int retry = 0;
+	do {
+		ret = recv_msg(sock, &fd, &err, sizeof(err));
+	} while (ret == 0 && retry++ < 3);
```
Comment (emersion) [inline backend/session/direct-ipc.c]: Do you think we could check for `errno` here? (Is it set to `EINTR` or something like that?)

**Comment by SirCmpwn** [comment]
Thanks!

**Comment by emersion** [comment]
@sghctoma Does FreeBSD still require `meson build -D b_lundef=false` as said in the README?

**Comment by sghctoma** [comment]
Yes, `b_lundef=false` is required.
However,  mesonbuild/meson@fd2c3b4c77ac1977d254301876525f8e631a940a fixed an issue that caused `override_options` to be uneffective for linker args. When that fix gets released, it will be possible to do something like this in `meson.build`:

```meson
if host_machine.system().startswith('freebsd')
  override_options = ['b_lundef=false']
else
  override_options = []
endif

lib_wlr = library(
  ...
  override_options: override_options,
)
```

**Comment by emersion** [comment]
Is this fix included in meson 0.47.2?

**Comment by sghctoma** [comment]
@emersion 

> Is this fix included in meson 0.47.2?
Nope, but it is in 0.48.0. Should I PR that change to meson.build now, or wait for 0.48.0 in the ports tree? BTW, my PR to update the port to 0.47.2 is still unanswered, I have no clue when it will be updated.

**Comment by emersion** [comment]
Sure, send a PR :)

**Pull request merged**